### PR TITLE
#168781821 add comment

### PR DESCRIPTION
**Pull request by Crispy-rw**
#### What does this PR do?
         Allows the owner of the request and a manager to comment on a request
#### Description of Task to be completed?
        This PR helps managers and client to have a discussion in a comment thread before 
        approving/rejecting s request
#### How should this be manually tested?
          - clone the repo
          - change branch to ft-add-comment-#168781821
          - run `_npm run db-migrate:dev_`
          - run `_npm run db-seeder:dev_`
          - run `_npm run dev_`

- Go in postman create a request 

> PUT localhost:3000/api/v1/requests/:requestId/comment

- Send in the body

> {"comment":"this is a comment"}

#### Any background context you want to provide?
          N/A
#### What are the relevant pivotal tracker stories?
          #168781821
#### Screenshots (if appropriate)
Screen Shot 2019-10-23 at 21 12 54

#### Questions:

### CODE REVIEW

File: src/controllers/comment.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import db from '../database/models/index';
+import sendResult from '../utils/sendResult';
+
+const comment = {
+  addComment: async (req, res) => {
+    const newComment = await db.Comments.create({
+      comment: req.body.comment,
+      userId: req.userData.userId,
+      requestId: 1,
```
Comment (lemoissonM) [inline src/controllers/comment.js]: the requestId will always be equal to 1?

**Review by lemoissonM — CHANGES_REQUESTED** [review]

File: src/database/migrations/20191022190917-add_comment.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.createTable('Comments',{
+      id: {allowNull: false,autoIncrement: true, primaryKey: true, type: Sequelize.INTEGER },
+      comment: {type:Sequelize.STRING, allowNull:false},
+      userId: {
+        type: 'integer',
+        allowNull: true,
+        defaultValue: 0,
```
Comment (lemoissonM) [inline src/database/migrations/20191022190917-add_comment.js]: Great job. However I don't think this should allow null, the default value is unnecessary in this case. there should be a user associated with any comment

File: src/database/migrations/20191022190917-add_comment.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        references: {
+          model: 'Users',
+          key: 'id',
+          onDelete: 'SET NULL'
+        }
+      },
+      requestId: {
+        type: 'integer',
+        allowNull: true,
+        defaultValue: 0,
```
Comment (lemoissonM) [inline src/database/migrations/20191022190917-add_comment.js]: The same here, allow null shouldn't be true

**Review by lemoissonM — CHANGES_REQUESTED** [review]

File: src/controllers/comment.js (hunk 1 above)
Comment (urbanishimwe) [reply src/controllers/comment.js]: Yup

File: src/database/models/add_comment.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+'use strict';
+module.exports = (sequelize, DataTypes) => {
+  const Comments = sequelize.define('Comments', {
+    comment: { type: DataTypes.STRING, allowNull: false },
+  }, { freezeTableName: true });
```
Comment (urbanishimwe) [inline src/database/models/add_comment.js]: There should be a functionalities for `replying to the comment`
like this:
```
parentCommentId: { type: DataTypes.Integer, allownull: true },
```
the, if `parentCommentId` is `null` will be able to identify that the comment was not a reply

File: src/middlewares/rolesMiddlewares.js [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
 import sendResult from '../utils/sendResult';
+import db from '../database/models/index';
 
 const notAuthorized = res => sendResult(res, 401, 'you are not authorized');
 const roles = {
   checkManager(req, res, next) {
     if (req.userData.role === 'manager') return next();
     return notAuthorized(res);
   },
-
```
Comment (urbanishimwe) [inline src/middlewares/rolesMiddlewares.js]: this space was necessary

File: src/database/migrations/20191022190917-add_comment.js (hunk 2 above)
Comment (urbanishimwe) [reply src/database/migrations/20191022190917-add_comment.js]: of course

File: src/middlewares/rolesMiddlewares.js [hunk 6]
```diff
@@ -16,6 +16,11 @@ const roles = {
   checkRequester(req, res, next) {
     if (req.userData.role === 'requester') return next();
     return notAuthorized(res);
-  }
+  },
+  async commentCheckUser(req, res, next) {
+    const Owner = await db.Requests.findOne({ where: { id: req.params.requestId }, attributes: ['UserId'] }, { raw: true });
+    if (req.userData.role === 'manager' || Owner.UserId === req.userData.userId) return next();
+    return notAuthorized(res);
```
Comment (urbanishimwe) [inline src/middlewares/rolesMiddlewares.js]: here any manager, who is not `the lineManger` of the user, can be able to comment!!, (`only the lineManager can comment to the request`);
i think it was suppossed to be like this:
```
if (req.userData.userId === lineManager || ........)
```
and after line 21, you can get the `lineManager` by writing:
```
const { lineManager } = await db.Users.findOne({ where: { id: Owner.UserId }, raw: true, attributes: ['lineManager'] });
```

File: src/validation/index.js [hunk 7]
```diff
@@ -85,8 +86,28 @@ const validator = {
       const message = (Object.keys(req.error).length === 0) ? err.message : req.error;
       return sendResult(res, 400, message);
     }
-  }
```
Comment (urbanishimwe) [inline src/validation/index.js]: space between functions, everywhere

File: src/middlewares/rolesMiddlewares.js (hunk 6 above)
Comment (Crispy-rw) [reply src/middlewares/rolesMiddlewares.js]: great idea but the pt board is saying the manager

Comment (urbanishimwe) [reply src/middlewares/rolesMiddlewares.js]: The PT board was trying to mean the manager of the requester

**Review by urbanishimwe — APPROVED** [review]

File: src/controllers/comment.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import db from '../database/models/index';
```
Comment (elchroy) [inline src/controllers/comment.js]: Change filename to `commentController.js` for consistency.

File: src/database/migrations/20191022190917-add_comment.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.createTable('Comments',{
+      id: {allowNull: false,autoIncrement: true, primaryKey: true, type: Sequelize.INTEGER },
```
Comment (elchroy) [inline src/database/migrations/20191022190917-add_comment.js]: Reformat these lines for consistency

File: src/database/migrations/20191022190917-add_comment.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        defaultValue: 0,
+        references: {
+          model: 'Requests',
+          key: 'id',
+        }
+      },
+      createdAt: { allowNull: false, type: Sequelize.DATE, defaultValue: Sequelize.NOW },
+      updatedAt: { allowNull: false, type: Sequelize.DATE, defaultValue: Sequelize.NOW },
+    })
+  },
```
Comment (elchroy) [inline src/database/migrations/20191022190917-add_comment.js]: These too!

File: src/database/migrations/20191022190917-add_comment.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.createTable('Comments',{
+      id: {allowNull: false,autoIncrement: true, primaryKey: true, type: Sequelize.INTEGER },
+      comment: {type:Sequelize.STRING, allowNull:false},
```
Comment (elchroy) [inline src/database/migrations/20191022190917-add_comment.js]: Reformat these lines for consistency

File: src/database/models/add_comment.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
+'use strict';
```
Comment (elchroy) [inline src/database/models/add_comment.js]: Change filename to `comment.js`

File: src/routes/requestRoutes.js [hunk 13]
```diff
@@ -5,19 +5,24 @@ import { validateTrips, validateAccommodation } from '../validation/trips';
...
-const { tripValidation } = valid;
+const { tripValidation, addCommentValidation, viewCommentValidation } = valid;
+const { addComment, viewComment } = comment;
 
 app.get('/', checkToken, checkRequester, requestController.getRequest);
 /* eslint-disable max-len */
 app.post('/', checkToken, checkRequester, valid.request, validateTrips, validateAccommodation, requestController.postRequest);
 app.get('/managers/:managerId', checkToken, checkManager, checkManagerId, getManagerRequests);
 app.patch('/:requestId/:status', checkToken, checkManager, checkExistingTrip, checkLineManager, tripValidation, changeRequestStatus);
+app.put('/:requestId/comment', addCommentValidation, checkToken, checkExistingTrip, checkTripOwner, addComment);
```
Comment (elchroy) [inline src/routes/requestRoutes.js]: Let's have a separate route for the comments. Kindly use POST to create comments.

File: src/tests/comment.spec.js [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+describe('/PUT comment', () => {
+  it('it should return 201 and create new comment data', (done) => {
+    chai.request(app)
+      .put('/api/v1/requests/1/comment')
+      .set('Authorization', helpers.createToken(5, 'admin@gmail.com', true, 'manager'))
+      .send({ comment: 'new Comment' })
+      .end((err, res) => {
+        expect(res.status).to.equal(201);
+        expect(res.body.status).to.equal(201);
```
Comment (elchroy) [inline src/tests/comment.spec.js]: Test that the res.body contains the same comment that was sent

File: src/tests/comment.spec.js [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+import app from '../index';
+import helpers from '../utils/helper';
+
+chai.use(chaiHttp);
+const { expect } = chai;
+
+describe('/PUT comment', () => {
+  it('it should return 201 and create new comment data', (done) => {
+    chai.request(app)
+      .put('/api/v1/requests/1/comment')
```
Comment (elchroy) [inline src/tests/comment.spec.js]: Change this to a POST request as well

**Pull request merged**